From ce489286c9f43f49f6c47f3be913bfac29ad1ddf Mon Sep 17 00:00:00 2001
From: Tianshu Qiu <tian.shu.qiu@intel.com>
Date: Fri, 5 Dec 2014 16:17:33 +0800
Subject: [PATCH 265/269] atomisp2: Unregister sub devices when removing
 atomisp2

This patch unregister the sub devices of atomisp2
before it's v4l2 device is unregistered.
In v4l2_device_unregister(), it unregister all sub
devices from system which make them not able to be
re-probed when reinstalling atomisp driver.
So we need to unregister or detach the sub devices
from atomisp v4l2 device before calling
v4l2_device_unregister().

Change-Id: I6b00ec2ad8e8fba3f7b92f9d9a9f10df2a48d2a9
Tracked-On:https://jira01.devtools.intel.com/browse/IMINAN-12256
Signed-off-by: Tianshu Qiu <tian.shu.qiu@intel.com>
Reviewed-on: https://icggerrit.ir.intel.com/31698
Integrator: Wei Tao <wei.tao@intel.com>
Reviewed-by: Jing Tao <jing.tao@intel.com>
Approver: Wei Tao <wei.tao@intel.com>
Build: Wei Tao <wei.tao@intel.com>
Reviewed-by: Wei Tao <wei.tao@intel.com>
Maintainer: Wei Tao <wei.tao@intel.com>
---
 .../drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c      | 8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
index 977688e..e9167c0 100644
--- a/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
+++ b/drivers/external_drivers/camera/drivers/media/pci/atomisp2/atomisp_driver/atomisp_v4l2.c
@@ -835,6 +835,9 @@ static int atomisp_subdev_probe(struct atomisp_device *isp)
 static void atomisp_unregister_entities(struct atomisp_device *isp)
 {
 	unsigned int i;
+#ifdef CONFIG_GMIN_INTEL_MID
+	struct v4l2_subdev *sd, *next;
+#endif
 
 	for (i = 0; i < isp->num_of_streams; i++)
 		atomisp_subdev_unregister_entities(&isp->asd[i]);
@@ -843,6 +846,11 @@ static void atomisp_unregister_entities(struct atomisp_device *isp)
 	for (i = 0; i < ATOMISP_CAMERA_NR_PORTS; i++)
 		atomisp_mipi_csi2_unregister_entities(&isp->csi2_port[i]);
 
+#ifdef CONFIG_GMIN_INTEL_MID
+	list_for_each_entry_safe(sd, next, &isp->v4l2_dev.subdevs, list)
+		v4l2_device_unregister_subdev(sd);
+#endif
+
 	v4l2_device_unregister(&isp->v4l2_dev);
 	media_device_unregister(&isp->media_dev);
 }
-- 
1.9.1

